File: examples/hooks/node/hook.js
#!/usr/bin/env node
// Small NetScope WebSocket hook examples: node hook.js <example> [ws://host:port]
const WebSocket = require("ws");
const fs = require("fs");

const examples = new Set(["log", "request-header", "remove-response-header", "rewrite-request-json",
  "rewrite-response-json", "block", "respond", "delay", "save", "observe"]);
const example = process.argv[2] || "log";
const url = process.argv[3] || "ws://127.0.0.1:8765";
if (!examples.has(example)) throw new Error(`choose one of: ${[...examples].join(", ")}`);
const role = example === "observe" ? "observe" : "intercept";
const bodyMode = example.startsWith("rewrite-") || example === "save" ? "json" : "none";

function decide(message) {
  const { event, phase } = message;
  const kind = event.kind;
  const headers = kind.headers || {};
  const host = headers.host || event.destination.host;
  const target = kind.target || "";
  const pattern = process.env.NETSCOPE_PATTERN || "blocked.example";
  if (example === "log") console.log(phase, kind.type, event.source, "->", event.destination);
  if (example === "request-header" && phase === "request_headers")
    return { action: "modify", set_headers: { "x-netscope-example": "node" } };
  if (example === "remove-response-header" && phase === "response_headers")
    return { action: "modify", remove_headers: ["server"] };
  if ((example === "rewrite-request-json" && phase === "request_body") ||
      (example === "rewrite-response-json" && phase === "response_body")) {
    const body = JSON.parse(Buffer.from(message.body_base64, "base64"));
    body.rewritten_by = "netscope-node";
    return { action: "modify", body_base64: Buffer.from(JSON.stringify(body)).toString("base64") };
  }
  if (example === "block" && phase === "request_headers" && (host.includes(pattern) || target.includes(pattern)))
    return { action: "drop", reason: "matched NETSCOPE_PATTERN" };
  if (example === "respond" && phase === "request_headers" && target.includes("/synthetic"))
    return { action: "respond", status: 200, headers: { "content-type": "application/json" },
      body_base64: Buffer.from('{"source":"netscope-node","synthetic":true}').toString("base64") };
  if (example === "delay" && (target.includes("/delay") || phase.endsWith("_body")))
    return { action: "delay", milliseconds: 500 };
  if (example === "save" && phase.endsWith("_body")) {
    fs.mkdirSync("bodies", { recursive: true });
    const path = `bodies/${message.request_id}-${phase}.bin`;
    fs.writeFileSync(path, Buffer.from(message.body_base64, "base64"));
    console.log(`saved ${path}`);
  }
  return { action: "continue" };
}

const socket = new WebSocket(url);
socket.on("open", () => socket.send(JSON.stringify({ type: "register", protocol_version: 1, role, body_mode: bodyMode })));
socket.on("message", raw => {
  try {
    const message = JSON.parse(raw);
    if (message.type === "registered") return console.log(message);
    if (message.type !== "event") return;
    const result = decide(message);
    if (role === "intercept") socket.send(JSON.stringify({ type: "action", request_id: message.request_id, ...result }));
  } catch (error) {
    console.error("hook error:", error.message);
  }
});
socket.on("error", error => console.error("socket error:", error.message));

File: examples/hooks/python/hook.py
#!/usr/bin/env python3
"""Small NetScope WebSocket hook examples: python hook.py <example> [ws://host:port]."""
import asyncio
import base64
import json
import os
import sys
from pathlib import Path

import websockets

EXAMPLES = {
    "log", "request-header", "remove-response-header", "rewrite-request-json",
    "rewrite-response-json", "block", "respond", "delay", "save", "observe",
}


def action(message, example):
    event = message["event"]
    kind = event["kind"]
    phase = message["phase"]
    headers = kind.get("headers", {})
    host = headers.get("host", event["destination"]["host"])
    target = kind.get("target", "")
    pattern = os.getenv("NETSCOPE_PATTERN", "blocked.example")

    if example == "log":
        print(phase, kind["type"], event["source"], "->", event["destination"], flush=True)
    elif example == "request-header" and phase == "request_headers":
        return {"action": "modify", "set_headers": {"x-netscope-example": "python"}}
    elif example == "remove-response-header" and phase == "response_headers":
        return {"action": "modify", "remove_headers": ["server"]}
    elif example in {"rewrite-request-json", "rewrite-response-json"} and phase == (
        "request_body" if example.startswith("rewrite-request") else "response_body"
    ):
        body = json.loads(base64.b64decode(message["body_base64"]))
        body["rewritten_by"] = "netscope-python"
        return {"action": "modify", "body_base64": base64.b64encode(json.dumps(body).encode()).decode()}
    elif example == "block" and phase == "request_headers" and (pattern in host or pattern in target):
        return {"action": "drop", "reason": "matched NETSCOPE_PATTERN"}
    elif example == "respond" and phase == "request_headers" and "/synthetic" in target:
        body = b'{"source":"netscope-python","synthetic":true}'
        return {"action": "respond", "status": 200, "headers": {"content-type": "application/json"},
                "body_base64": base64.b64encode(body).decode()}
    elif example == "delay" and ("/delay" in target or phase.endswith("_body")):
        return {"action": "delay", "milliseconds": 500}
    elif example == "save" and phase.endswith("_body"):
        directory = Path("bodies")
        directory.mkdir(exist_ok=True)
        path = directory / f'{message["request_id"]}-{phase}.bin'
        path.write_bytes(base64.b64decode(message["body_base64"]))
        print(f"saved {path}", flush=True)
    return {"action": "continue"}


async def main():
    example = sys.argv[1] if len(sys.argv) > 1 else "log"
    if example not in EXAMPLES:
        raise SystemExit(f"choose one of: {', '.join(sorted(EXAMPLES))}")
    url = sys.argv[2] if len(sys.argv) > 2 else "ws://127.0.0.1:8765"
    role = "observe" if example == "observe" else "intercept"
    body_mode = "json" if example.startswith("rewrite-") or example == "save" else "none"
    # NetScope v1 only reads an active hook socket while awaiting an action. Disable client-side
    # keepalive pings so an idle example stays connected between traffic events.
    async with websockets.connect(url, ping_interval=None) as socket:
        await socket.send(json.dumps({"type": "register", "protocol_version": 1,
                                      "role": role, "body_mode": body_mode}))
        print(await socket.recv())
        async for raw in socket:
            message = None
            try:
                message = json.loads(raw)
                if message.get("type") != "event":
                    continue
                result = action(message, example)
                if role == "intercept":
                    result.update({"type": "action", "request_id": message["request_id"]})
                    await socket.send(json.dumps(result))
            except Exception as error:
                print(f"hook error: {error}", file=sys.stderr, flush=True)
                if role == "intercept" and message is not None:
                    await socket.send(json.dumps({"type": "action", "request_id": message["request_id"], "action": "continue"}))


if __name__ == "__main__":
    asyncio.run(main())
